crates/red_knot_python_semantic/src/semantic_index/constraint.rs

```diff
@@ -7,7 +7,13 @@ use crate::semantic_index::expression::Expression;
 use crate::semantic_index::symbol::{FileScopeId, ScopeId};
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub(crate) enum Constraint<'db> {
+pub(crate) struct Constraint<'db> {
+    pub(crate) node: ConstraintNode<'db>,
+    pub(crate) negative: bool,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub(crate) enum ConstraintNode<'db> {
```

Comment:
In this context, "constraint" term is used for both the node that applies the constraint, and for the outcome of the narrowing process. 
What do you think about renaming this one into `TypeGuard` and use `constraint` name only for the output (map from symbol ids to types)?

Reply:
Hmm. I'm open to a renaming here to better distinguish these, but not sure I like `TypeGuard`, I'd rather avoid that term outside of dealing with the actual `typing.TypeGuard` feature.

Reply:
We often called these types *Kind* in other places.

Reply:
What do you think about "Predicate"? I like that it's more easy to understand what "negative" means on a predicate, than on Type or Kind as of the negation of a predicate doesn't always imply the negation of its outcome constraint.

Reply:
I'm happy with "Predicate"!